collectors/modality-probe-collector-common/src/lib.rs

```diff
@@ -283,320 +354,414 @@ impl TryFrom<&LogFileRow> for LogEntry {
-                data: LogEntryData::LogicalClock(
-                    clock_bucket.id,
-                    clock_bucket.epoch,
-                    clock_bucket.ticks,
-                ),
-                receive_time,
-            });
+    for fc in &log_report.frontier_clocks {
+        log_entries_buffer.push(ReportLogEntry {
+            session_id,
+            sequence_number,
+            sequence_index,
+            probe_id,
+            data: LogEntryData::FrontierClock(*fc),
+            receive_time,
+        });
+        sequence_index += 1;
+    }
 
-            segment_index += 1;
+    for event in &log_report.event_log {
+        log_entries_buffer.push(ReportLogEntry {
+            session_id,
+            sequence_number,
+            sequence_index,
+            probe_id,
+            data: LogEntryData::from(*event),
+            receive_time,
+        });
+        sequence_index += 1;
+    }
+}
+
+impl TryFrom<&[u8]> for Report {
+    type Error = SerializationError;
+    fn try_from(buf: &[u8]) -> Result<Self, Self::Error> {
+        let report = WireReport::new(buf)?;
+        let (epoch, ticks) = modality_probe::unpack_clock_word(report.clock());
+        let id = report.probe_id()?;
+        let mut owned_report = Report {
+            probe_id: id,
+            probe_clock: LogicalClock { id, epoch, ticks },
+            seq_num: report.seq_num(),
+            frontier_clocks: vec![],
+            event_log: vec![],
+        };
+
+        let clocks_len = report.n_clocks() as usize * mem::size_of::<LogicalClock>();
+        let mut probe_id = None;
+        let payload = report.payload();
+        for u32_bytes in payload[..clocks_len].chunks_exact(mem::size_of::<LogEntry>()) {
+            let raw = le_bytes::read_u32(u32_bytes);
+            if probe_id.is_none() {
+                probe_id = ProbeId::new(
+                    unsafe { LogEntry::new_unchecked(raw) }.interpret_as_logical_clock_probe_id(),
+                );
+            } else {
+                let (epoch, ticks) = modality_probe::unpack_clock_word(raw);
+                owned_report.frontier_clocks.push(LogicalClock {
+                    id: probe_id.expect("checked above that probe id is not none"),
+                    epoch,
+                    ticks,
+                });
+                probe_id = None;
+            }
         }
 
-        for event in &segment.events {
-            match event {
-                LogEvent::Event(ev) => {
-                    log_entries_buffer.push(LogEntry {
-                        session_id,
-                        segment_id: next_segment_id.into(),
-                        segment_index,
-                        probe_id,
-                        data: LogEntryData::Event(*ev),
-                        receive_time,
-                    });
+        let mut interpret_next_as = Next::DontKnow;
+        for u32_bytes in payload[clocks_len..].chunks_exact(mem::size_of::<LogEntry>()) {
```

Comment:
This interpretation code probably has a canonical place to live somewhere in our log module (or, if there's divergence now, in the reporting module).  In the previous version of this I wrote some iterators in this vein to prevent rework across collector implementations, and would recommend the same here, since it seems probably unnecessary to require collectors to know the deep details of the log-format.

Reply:
Now that I look up and see that this is in a cross-collector lib crate, the above comment seems mostly superfluous.